src/crypto/NoteEncryption.cpp

```diff
@@ -0,0 +1,161 @@
+}
+
+template<size_t MLEN>
+typename NoteEncryption<MLEN>::Ciphertext NoteEncryption<MLEN>::encrypt
+                                          (const uint256 &pk_enc,
+                                           const NoteEncryption<MLEN>::Plaintext &message
+                                          )
+{
+    uint256 dhsecret;
+
+    if (crypto_scalarmult(dhsecret.begin(), esk.begin(), pk_enc.begin()) != 0) {
+        throw std::logic_error("Could not create DH secret");
+    }
+
+    // Construct the symmetric key
+    unsigned char K[NOTEENCRYPTION_CIPHER_KEYSIZE];
+    KDF(K, dhsecret, epk, pk_enc, seed, nonce);
+
+    // Increment the number of encryptions we've performed
+    nonce++;
+
+    // The nonce is zero for our purposes
+    unsigned char cipher_nonce[crypto_aead_chacha20poly1305_NPUBBYTES] = {};
+
+    NoteEncryption<MLEN>::Ciphertext ciphertext;
+
+    crypto_aead_chacha20poly1305_encrypt(ciphertext.begin(), NULL,
+                                         message.begin(), MLEN,
+                                         NULL, 0, // no "additional data"
+                                         NULL, cipher_nonce, K);
+
+    return ciphertext;
+}
+
+template<size_t MLEN>
+typename NoteEncryption<MLEN>::Plaintext NoteEncryption<MLEN>::decrypt
+                                         (const uint256 &sk_enc,
+                                          const NoteEncryption<MLEN>::Ciphertext &ciphertext,
+                                          const uint256 &epk,
+                                          const uint256 &seed,
+                                          unsigned char nonce
+                                         )
+{
+    uint256 pk_enc = generate_pubkey(sk_enc);
+
+    uint256 dhsecret;
+
+    if (crypto_scalarmult(dhsecret.begin(), sk_enc.begin(), epk.begin()) != 0) {
+        throw std::logic_error("Could not create DH secret");
+    }
+
+    unsigned char K[NOTEENCRYPTION_CIPHER_KEYSIZE];
+    KDF(K, dhsecret, epk, pk_enc, seed, nonce);
+
+    // The nonce is zero for our purposes
+    unsigned char cipher_nonce[crypto_aead_chacha20poly1305_NPUBBYTES] = {};
+
+    NoteEncryption<MLEN>::Plaintext plaintext;
+
+    if (crypto_aead_chacha20poly1305_decrypt(plaintext.begin(), NULL,
+                                             NULL,
+                                             ciphertext.begin(), NoteEncryption<MLEN>::CLEN,
+                                             NULL,
+                                             0,
+                                             cipher_nonce, K) != 0) {
+        throw std::runtime_error("Could not decrypt message");
+    }
+
+    return plaintext;
+}
+
+template<size_t MLEN>
+uint256 NoteEncryption<MLEN>::generate_privkey(const uint256 &a_sk)
+{
+    uint256 sk = PRF_addr<1>(a_sk);
+
+    clamp_curve25519(sk.begin());
+
+    return sk;
+}
+
+template<size_t MLEN>
+uint256 NoteEncryption<MLEN>::generate_pubkey(const uint256 &sk_enc)
+{
```

Comment:
I do with `esk`, which doesn't need to be clamped. So more accurately, nothing is assumed to be clamped except `sk_enc`, which is always clamped since it's always created by `generate_privkey`.